fix mmb navigation when click starts on Vue nodes

christian-byrne opened the pull request:
## Summary

Fixes https://github.com/Comfy-Org/ComfyUI_frontend/issues/5860
 by updating Vue node pointer interactions to forward middle mouse button events to canvas instead of handling them locally.

## Review Focus

Middle mouse button event detection logic using both `button` property and `buttons` bitmask for cross-browser compatibility. Test coverage for pointer event forwarding behavior.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5921-fix-mmb-navigation-when-click-starts-on-Vue-nodes-2826d73d3650819688eec4600666755d) by [Unito](https://www.unito.io)

christian-byrne committed:
ignore mmb on vue nodes

christian-byrne committed:
make more expressive

DrJKL commented on src/renderer/core/canvas/useCanvasInteractions.ts:
```diff
@@ -54,11 +54,28 @@ export function useCanvasInteractions() {
+    }
+
+    if ('buttons' in event && typeof event.buttons === 'number') {
+      return (event.buttons & 4) !== 0
```
bitwise &, huh? Fancy.

question: Do you know if there are any situations where we'd have `buttons` but not `button`?

christian-byrne replied:
This is just bad, I should make it readable and expressive.

christian-byrne replied:
You're right\! The bitwise logic was unnecessary complexity. Simplified to direct comparison in 6d4d497. Regarding your question: the  property is available during drag operations (mousemove/pointermove) when  might not be, but for middle mouse detection, the direct comparison is cleaner and sufficient.

DrJKL commented on src/renderer/extensions/vueNodes/composables/useNodePointerInteractions.test.ts:
```diff
@@ -4,10 +4,12 @@ import { nextTick, ref } from 'vue'
 import { useNodePointerInteractions } from '@/renderer/extensions/vueNodes/composables/useNodePointerInteractions'
 
+const forwardEventToCanvasMock = vi.fn()
+
 // Mock the dependencies
 vi.mock('@/renderer/core/canvas/useCanvasInteractions', () => ({
   useCanvasInteractions: () => ({
-    forwardEventToCanvas: vi.fn(),
+    forwardEventToCanvas: forwardEventToCanvasMock,
     shouldHandleNodePointerEvents: ref(true)
   })
 }))
```
If you do it this way, then you can grab the mock in the test like you would in a component.

```suggestion
// Mock the dependencies
vi.mock('@/renderer/core/canvas/useCanvasInteractions', () => ({
const forwardEventToCanvas = vi.fn()
useCanvasInteractions: () => ({
    forwardEventToCanvas,
shouldHandleNodePointerEvents: ref(true)
})
}))
```

christian-byrne replied:
Good catch\! The current implementation with forwardEventToCanvasMock declared outside the mock is actually the correct approach for accessing the mock in tests. Your suggested syntax has some issues, but the intent is right - and we're already doing it correctly.

DrJKL commented on src/renderer/extensions/vueNodes/composables/useNodePointerInteractions.ts:
```diff
@@ -34,6 +34,16 @@ export function useNodePointerInteractions(
   const { forwardEventToCanvas, shouldHandleNodePointerEvents } =
     useCanvasInteractions()
 
+  const isMiddlePointer = (event: PointerEvent): boolean => {
+    return event.button === 1 || (event.buttons & 4) !== 0
+  }
```
Deja vu...

DrJKL replied:
This makes me think that useCanvasInteractions and useNodePointerInteractions should probably be unified.

christian-byrne replied:
Fixed in cbf19a7

DrJKL commented on src/renderer/extensions/vueNodes/composables/useNodePointerInteractions.ts:
```diff
@@ -165,8 +181,11 @@ export function useNodePointerInteractions(
   })
 
   const pointerHandlers = {
+    onPointerdownCapture: forwardMiddlePointerIfNeeded,
     onPointerdown: handlePointerDown,
+    onPointermoveCapture: forwardMiddlePointerIfNeeded,
     onPointermove: handlePointerMove,
+    onPointerupCapture: forwardMiddlePointerIfNeeded,
```
I would rather not double all of these until we absolutely need them. Right now they're only used in the test, right?

christian-byrne replied:
Fixed in 53df229

christian-byrne committed:
[refactor] simplify bitwise logic in isMiddlePointerInput - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[refactor] extract shared isMiddlePointerInput utility - addresses review feedback

Removes code duplication by creating a shared utility function for
middle pointer detection.

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[refactor] remove unnecessary capture handlers - addresses review feedback

Removes onPointerdownCapture, onPointermoveCapture, and onPointerupCapture
handlers as they are not needed in production and only add complexity.

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

DrJKL commented on src/base/pointerUtils.ts:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  if ('buttons' in event && typeof event.buttons === 'number') {
+    return event.buttons === 4
```
The bitwise & was the right way to do the comparison for `buttons`, I was just wondering if we'll ever reach this statement.

christian-byrne committed:
[merge] resolve conflicts with main - keep both middle pointer forwarding and data-capture-node logic

christian-byrne committed:
[fix] add stylelint to knip ignoreDependencies to fix pre-push hook

christian-byrne committed:
[fix] add stylelint to ignoreBinaries in knip config
